pep-9999.rst

```diff
@@ -0,0 +1,173 @@
+PEP: 9999
+Title: Format Specifier for Signed Zero
+Author: John Belmonte <john@neggie.net>
+Sponsor: Mark Dickinson
+Discussions-To: https://bugs.python.org/issue45995
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 29-Jan-2022
+Python-Version: 3.11
+Post-History:
+
+
+Abstract
+========
+
+Though float and Decimal types can represent `signed zero`_, in many fields
+of mathematics negative zero is surprising or unwanted-- especially in the
+in the context of displaying an (often rounded) numerical result.  This PEP
+proposes an extension to the string format specification, allowing negative
+zero to be normalized to positive zero.
+
+.. _`signed zero`:  https://en.wikipedia.org/wiki/Signed_zero
+
+
+Background
+==========
+
+Here is negative zero::
+
+    >>> x = -0.
+    >>> x
+    -0.0
+
+When formatting a number, negative zero can result from rounding.  Assuming
+the user's intention is truly to discard precision, the distinction between
+negative and positive zero of the rounded result might be considered an
+unwanted artifact::
+
+    >>> for x in (.002, -.001, .060):
+    ...     print(f'{x: .1f})
+     0.0
+    -0.0
+     0.1
+
+There are various approaches to clearing the sign of a negative zero.  It
+can be achieved without a conditional by adding positive zero::
+
+    >>> x = -0.
+    >>> x + 0.
+    0.0
+
+To normalize negative zero when formatting, it would be necessary to perform
+a redundant (and error-prone) pre-rounding of the input::
+
+    >>> for x in (.002, -.001, .060):
+    ...     print(f'{round(x, 1) + 0.: .1f}')
+     0.0
+     0.0
+     0.1
+
+The behavior of number formatting with regard to negative zero is not
+unique to Python.  The first low-level implementations of ``printf()``
+and such likely wired the sign bit to ``-`` in the output without special
+consideration of the zero mantissa case.  In hindsight, it would have
+been nice to have a formatting option for treatment of negative zero--
+perhaps even coercing it to positive zero by default.  However, for better
+or worse, existing number formatting implementations inherit from the
+original specimens, passing the zero sign through without option.
+
+
+Proposal
+========
+
+An optional, literal ``z`` is added to the
+`Format Specification Mini-Language`_ following ``sign``::
+
+    [[fill]align][sign][z][#][0][width][grouping_option][.precision][type]
+
+where ``z`` is allowed only for float and Decimal types.  The new specifier
+is supported by both f-strings and ``str.format()``, but not %-formatting.
+
```

Comment:
It would be more correct to say that support of 'z' is provided by `float.__format__`, `decimal.__format__`, and I guess `complex.__format__`. A consequence of that is that only f-strings, built-in format(), and str.format() will allow 'z' to be used, and not %-formatting.

Reply:
I guess `int.__format__` too?   Even though there isn't an int -0, we don't want an exception when someone hands an int to a format like `z.1f`.

Reply:
Hmm. Good question. I believe that the `f` forces the int to be converted to a float, so that `z.1f` would automatically be supported for ints.

Reply:
Couldn't this be tested and confirmed on your reference implementation?

Reply:
> I believe that the f forces the int to be converted to a float, so that z.1f would automatically be supported for ints.

Indeed, this is correct.  I was confusing it with having a integer type in the format string.  The implementation behaves as expected:
```python
>>> f"{0:z.1f}"
'0.0'
>>> f"{0:zd}"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Negative zero coercion (z) not allowed in integer format specifier
```